Cleanups and workflow improvements

LecrisUT opened the pull request:
For now I have kept this PR very minimal to just do some minor cleanup and address a packaging issue:
- Drop the setup.py file. Not needed with the PEP621 format, plus you are not using `setuptools` as the build backend anymore
- Make sure tests and doc sources are present in the sdist for downstream packaging. I would even recommend to just not do any filtering. sdist size is really not an issue for non-compiled projects
- Added a basic pre-commit. Did not hook it to the github actions since you probably want a more dedicated way of managing these and I don't want to step on your toes with my organization opinions on that

A few other things that could be added, but I still leave it open for your approval:
- ruff as a linter (with many checks turned off for now)
- uv.lock and renovate workflow so that we are always testing the latest releases (would like a more agnostic pylock.toml but there are no tools that use it as the main source yet)
- migrating the dependencies to dependecy-groups (obsoleting the `requirements.txt`) and splitting it in more granular dependencies so that it can be reused by downstream, at least `test`, `docs`, `dev`
- drop the github.io reference in favor of RTD?

jnikula commented:
> For now I have kept this PR very minimal to just do some minor cleanup and address a packaging issue:

Please add the commit messages to individual commits. I use the rebase and merge strategy, and the github pull request text is not recorded in git history. Only the individual commit messages are.

> * Drop the setup.py file. Not needed with the PEP621 format, plus you are not using `setuptools` as the build backend anymore

Hmm, I think this used to be required for editable installs. Wonder when that changed. My development workflow seems to work without. Good riddance.

> * Make sure tests and doc sources are present in the sdist for downstream packaging. I would even recommend to just not do any filtering. sdist size is really not an issue for non-compiled projects

I don't really have strong opinions or expertise on this.

> * Added a basic pre-commit. Did not hook it to the github actions since you probably want a more dedicated way of managing these and I don't want to step on your toes with my organization opinions on that

Personally, I've never been keen on pre-commit hooks, and also not keen on having them depend on github setup. I'm not sure what to make of this.

My style has been to add Makefile targets for static checks, with `make check` running all of them. You can then do this both locally and in the github actions.

I guess I'd need more rationale for this.

> A few other things that could be added, but I still leave it open for your approval:
> 
> * ruff as a linter (with many checks turned off for now)

I've experimented with it, with the idea of switching to it completely, throwing flake8 in the curb. Seems like a nice tool, and would get rid of `setup.cfg` as flake8 can't seem to get `pyproject.toml` support done.

> * uv.lock and renovate workflow so that we are always testing the latest releases (would like a more agnostic pylock.toml but there are no tools that use it as the main source yet)

Please elaborate. Please note that pretty much all of my experience in Python packaging is with this project, largely on my own, and very little outside influence. I just might have huge gaps in my knowledge and never even know it.

> * migrating the dependencies to dependecy-groups (obsoleting the `requirements.txt`) and splitting it in more granular dependencies so that it can be reused by downstream, at least `test`, `docs`, `dev`

I only ever use `requirements.txt` for testing and development. `pyproject.toml` has the optional dependencies, but I don't mind making that more granular.

> * drop the github.io reference in favor of RTD?

I've been tempted to do the opposite, and just remove the project from RTD actually. I get the github.io builds with the github actions, with direct and immediate feedback if something goes wrong. RTD is more detached from the setup.

LecrisUT committed:
Add `uv.lock`/`pylock.toml` files

These are not yet being used in the CI or Makefile. More design work is needed

Signed-off-by: Cristian Le <git@lecris.dev>

LecrisUT committed:
Add a `uv` workflow in the venv if possible

Signed-off-by: Cristian Le <git@lecris.dev>

LecrisUT committed:
Use the `uv.lock` dependencies in the CI

Signed-off-by: Cristian Le <git@lecris.dev>

jnikula commented:
I understand adding `.pre-commit-config.yaml` as something like communication of the standards being used. A bit like `.editorconfig`. You don't have to use it, but you can if you like. And, frankly, I don't think I personally would. I dislike the idea of letting it install everything from git repositories, and maintaining its own environments and caches of installed things separately from everything else. I wouldn't want to run that manually either.

The question then becomes, should pre-commit be a dependency of the project? Should it be used in CI? If I'm not going to use it myself, I'm going to maintain being able to run all the checkers with `make check`, and I don't want those to depend on pre-commit either.

---

I think I'd like more rationale for `uv.lock` and `pylock.toml`. In general, everything should work with the latest packages, users are going to get the latest packages from pypi anyway, why should we limit to a fixed set?

LecrisUT commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,20 @@
+# SPDX-FileCopyrightText: 2025 Jani Nikula <jani@nikula.org>
```
Nothing particular, it's just not something particularly copyrightable/worth copyrighting. My contributions are generally more visible in commit histories where the "copyright" is more ephemeral and vanishing so that when no code lines that I wrote survived so to would my "copyright" be irrelevant.

But that's just philosophy, in practice I confirm to the preferences of the maintainers.

LecrisUT commented:
Regarding the maintenance problem of pre-commit, that is a valid issue. If it becomes a maintenance burden that you are not ready to take on and the tradeoff of occasionally having to do the formatting manually for the PR author is a better tradeoff then probably it is fine to skip it. Instead maybe some badges of the tooling used may help.

LecrisUT committed:
Add makefile targets and CI for uv.lock

Signed-off-by: Cristian Le <git@lecris.dev>

jnikula commented:
Apologies for the long delay. Overall I think I'm ready to jump on board with `uv`.

First, I think we should address this issue:

```
$ uv lock
warning: The `requires-python` specifier (`~=3.9`) in `hawkmoth` uses the tilde specifier (`~=`) without a patch version. This will be interpreted as `>=3.9, <4`. Did you mean `~=3.9.0` to constrain the version as `>=3.9.0, <3.10`? We recommend only using the tilde specifier with a patch version to avoid ambiguity.
```

I believe it would be covered with using `requires-python = ">=3.9, <4"`.

I'll comment on some specific items in the commits separately.

jnikula commented on venv:
```diff
@@ -1,8 +1,15 @@
-. ${1:-.venv}/bin/activate
-pip install --upgrade pip
-pip install --editable . --group dev
+if command -v uv; then
```
I think we should embrace uv fully here, and just drop the conditional stuff. IIRC @BrunoMSantos doesn't use the `venv` script anyway. Maybe we could drop the entire script?

Can we run the checkers and tests using uv both locally and in CI, without explicitly entering the virtual environment? It seems to me that's the recommended uv way anyway.

Is it possible to configure uv to use `--system-site-packages` for anything that needs libclang?

BrunoMSantos replied:
I do use venvs sometimes to check that something works with different versions of this and that, but uv looks increasingly like the new default way of managing them. So :+1:

I've only recently started using uv myself though, I'll defer on the system-site-packages issue.

LecrisUT replied:
> Can we run the checkers and tests using uv both locally and in CI, without explicitly entering the virtual environment? It seems to me that's the recommended uv way anyway.

It's a case-by-case basis. `optional-groups` usually yes, but you still need a step to install in the local venv. For `mypy` it is recommended to use the runtime dependencies and entering a venv. I can make it a case-by-case for the actions, or I can unify it with a single `dev`, your call.

jnikula commented on .github/workflows/makefile.yml:
```diff
@@ -28,7 +31,10 @@ jobs:
+      - uses: astral-sh/setup-uv@v7
+        with:
+          activate-environment: true
+      - run: uv pip install --group dev
```
Ditto here, and below.

LecrisUT commented:
Ok, I am at the PC now, and I'll look into the comments. Because of Github review I have to reply one by one, so first off:

> I believe it would be covered with using `requires-python = ">=3.9, <4"`.

It is a rather minor warning IMO, but yeah up to you if you want `>=3.9` or `>=3.9, <4`. The former is the more common because all non-compiled wheels are already locked by ABI3 so the higher bound is already implied. The only effect would be in the sdist, for which I usually leave it to the users to discover and report the problem rather than prematurely block them. Generally it is discouraged to higher bound dependencies especially in low maintenance projects because that usually creates time pressure and more maintenance burden when the newer versions are out.

LecrisUT commented on .github/workflows/makefile.yml:
```diff
@@ -18,7 +18,10 @@ jobs:
+          activate-environment: true
+      - run: uv pip install --group dev
       - name: Style check
         run: make check-style
```
> Here too, can we run the checks using uv instead of make

For sure and would recommend it. There are 3 ways to call, and I do not have explicit preference:
- Use the `ruff` et. al. commands directly
- Use PEP723 style scripts (basically the same as a shell script but includes headers for dependencies). I am not sure if it works with dependency-groups though
- Use yet another task manager that is python native e.g. `tox` or `nox`

I thought there was also a `uv` specific task manager like `hatch` has, but could not find it.

LecrisUT commented on Makefile.local:
```diff
@@ -8,6 +8,17 @@ RST := $(RST) CHANGELOG.rst CONTRIBUTING.rst README.rst
+.PHONY: check-lockfile
+check-lockfile:
+	uv lock
+	uv export --frozen --format=pylock.toml --output-file=pylock.toml --all-extras --all-groups --quiet
```
Fair point, I will update it and stage this. Will push the changes when the other points are resolved
